# **HW1: Data Analysis and Visualization with Pandas and Matplotlib**

#### **Attention:** *This assignment is to be completed individually. Group work is not permitted.*


### **TODO: Please Write Your Full Name here**

## **NAME: Jonathan Cruz**



## **Getting the Dataset for the Homework**

* [University of California, Irvine Machine Learning Repository](https://archive.ics.uci.edu/) is a popular resource for many classic datasets used in machine learning and statistics. It contains a number of freely available datasets for learners to practice and improve their data analysis and machine learning skills or conduct experiments for their reseach. We will use one of the datasets from this repository for HW1.

**Dataset Citation:** Schlimmer,Jeffrey. (1987). Automobile. UCI Machine
Learning Repository. https://doi.org/10.24432/C5B01C.


### **Q1: Load the Dataset into a Pandas DataFrame**


We will need a pandas dataframe to be able to analyze the dataset that is in the CSV file. Let's first `import` the Pandas library.

In [1]:
import pandas as pd

* If you open (double click) the `automobile.csv` file you downloaded, you will notice that it lacks a header (column names). To address this, we can utilize the `names` parameter from the `read_csv` function in pandas. This parameter enables us to assign column names to each data point separated by commas, effectively treating them as distinct columns. But we first need to define the column names. We get the column names for this dataset based on the documentation. See "Attribute Information": https://archive.ics.uci.edu/dataset/10/automobile



In [2]:
# Create a python "list" for column names.
column_names = [
    "symboling", "normalized-losses", "make", "fuel-type", "aspiration",
    "num-of-doors", "body-style", "drive-wheels", "engine-location", "wheel-base",
    "length", "width", "height", "curb-weight", "engine-type",
    "num-of-cylinders", "engine-size", "fuel-system", "bore", "stroke",
    "compression-ratio", "horsepower", "peak-rpm", "city-mpg", "highway-mpg", "price"
]

* We are now ready to create the dataframe using the `read_csv()` function from pandas library. But there is one more thing we will have to be careful about. When we opened the file, we noticed that there were rows an columns where a `?` was placed. This is a common practice for dataset creators to fill the missing values with some characters or strings such as `?`, '999`, 'NA`, etc. This is also why when analyzing datasets, it's essential to refer to the dataset's documentation or metadata (if available) to understand how missing values are represented.

* Since we identified how this specific dataset represents missing values, we can use a very handy parameter from `read_csv` function, `na_values` which specifies additional strings to recognize as NaN (Not a Number), which means missing or null values. `na_values="?"` argument indicates that any occurrence of the "?" string in the dataset should be treated as a NaN value. This parameter will allow us to correctly process our dataset and we handle the missing values appropriately.



Now with all this new knowledge you learned about `na_values` , let's create the dataframe.


In [3]:
#TODO: Write your code here
# Create the dataframe, using pandas `read_csv` by using the na_values parameter discussed above.
df = pd.read_csv("automobile (1).csv", na_values="?")

#Display the first few rows
df.head()


,3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,...,130,mpfi,3.47,2.68,9.00,111,5000,21,27,13495
0,3,NaN,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111.0,5000.0,21,27,16500.0
1,1,NaN,alfa-romero,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154.0,5000.0,19,26,16500.0
2,2,164.0,audi,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102.0,5500.0,24,30,13950.0
3,2,164.0,audi,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115.0,5500.0,18,22,17450.0
4,2,NaN,audi,gas,std,two,sedan,fwd,front,99.8,...,136,mpfi,3.19,3.40,8.5,110.0,5500.0,19,25,15250.0


### **Q2: Explore the Dataset**

**2.1 Get the shape of the dataset**




In [4]:
# TODO: Write your code here
df.shape

(204, 26)

**2.2 Check for missing values: Sum how many missing values are in each column**

*Note: If we did not use `na_values`, we would get a 0 for missing values for all rows, as pandas read_csv would interpret the `?` symbol as `object`, i.e. a string.*

In [6]:
# TODO: Write your code here
df.isna().sum()

3               0
?              40
alfa-romero     0
gas             0
std             0
two             2
convertible     0
rwd             0
front           0
88.60           0
168.80          0
64.10           0
48.80           0
2548            0
dohc            0
four            0
130             0
mpfi            0
3.47            4
2.68            4
9.00            0
111             2
5000            2
21              0
27              0
13495           4
dtype: int64

**2.3 What are the data types in each column?**

In [7]:
# TODO: Write your code here
df.dtypes

3                int64
?              float64
alfa-romero     object
gas             object
std             object
two             object
convertible     object
rwd             object
front           object
88.60          float64
168.80         float64
64.10          float64
48.80          float64
2548             int64
dohc            object
four            object
130              int64
mpfi            object
3.47           float64
2.68           float64
9.00           float64
111            float64
5000           float64
21               int64
27               int64
13495          float64
dtype: object

**2.4 Filter cars with a price above the average price**
* Store your 'subset' with filtered cars whose price is above the average of all cars in a dataframe named `above_the_average_price` and return the first few rows from it.

In [8]:
# TODO: Write your code here
avg_price = df["price"].mean()
above_the_average_price = df[df["price"] > avg_price]
above_the_average_price.head()

KeyError: 'price'

**2.5 Group the cars by `make` and get the average `price` for each make.**

Store them in a dataframe named `average_price_by_make` and sort them in descending order.

In [ ]:
#TODO: Write your code here

### **Q3 Visualize the Dataset Using Matplotlib**


**3.1: Plot a histogram of the `horsepower` column.**

Using the given dataset, create a histogram to visualize the distribution of the `horsepower` column. Follow the guidelines below to ensure your visualization is informative:

* Drop any missing values from the horsepower column before plotting.
* Use 20 bins for your histogram.
* Set the color of the bars to any color you like with an `alpha` (transparency) of 0.7.
* The figure size should be 15 units in width and 6 units in height.
* The title of the plot should be "Distribution of Horsepower".
* Label the x-axis as "Horsepower" and the y-axis as "Number of Cars".
* (Optional) For better readability, rotate the x-axis labels by 45 degrees.
Display a grid in the plot for clarity.

*Tip: If your chart appears cluttered, consider adjusting the figure size to at least 15 x 6 or more for better clarity and readability.*

In [ ]:
import matplotlib.pyplot as plt

#TODO: Write your code here

**3.2: Plot a bar chart of the `make` column, showing the count of each make.**

Using the dataset, create a bar chart to visualize the number of cars for each manufacturer (or "make") in the dataset. Follow the guidelines below to ensure your visualization is informative:

* Compute the value counts of the make column to determine the number of cars for each manufacturer.
* Use a figure size of 15 units in width and 6 units in height for your plot.
* Set the color of the bars to purple with an alpha (transparency) of 0.7.
* The title of your plot should be "Number of Cars by Make".
* Label the x-axis as "Make" and the y-axis as "Number of Cars".
* (Optional) For better readability, rotate the x-axis labels by 45 degrees.
* Display a grid along the y-axis to aid in readability.




pandas [value_counts](https://www.educative.io/answers/what-is-valuecounts-in-pandas) function could be useful here.

In [ ]:
# TODO: Write your code here

**3.3: Plot a scatter plot between price and horsepower.**

Using the provided dataset, create a scatter plot to visualize the relationship between the horsepower of a car and its price. Follow the guidelines below to ensure your visualization is informative:

* Use the horsepower column values for the x-axis and the price column values for the y-axis.
* Set the alpha (transparency) of the data points to 0.6.
* The title of your plot should be "Scatter plot between Price and Horsepower".
* Label the x-axis as "Horsepower" and the y-axis as "Price".
* Enable the grid on your plot for clarity.


In [ ]:
# TODO: Write your code here

**3.4 Based on the scatter plot, histogram, and bar chart you've visualized, can you describe any patterns, trends, or insights you observe in the data?**

**TODO:** *Write your answer to the question #3.4 here in 2-3 sentences for each plot you have visualized above.*

Some questions you can think of while writing up your thoughts:

* Which range of horsepower is the most common among the cars in the dataset?
* Which manufacturer has the highest number of cars in the dataset?
* Is there any noticeable relationship between the horsepower and the price of the cars? If yes, describe it.
* Are there any "outliers" or "clusters" of data points that stand out?